### DISCUSSION TITLE
How to propagate Airflow's log level to KubernetesPodOperator?

### DISCUSSION THREAD

**Discussion by ketozhang** [apache/airflow#38083] [Q&A]
KPO doesn't share state with the rest of the DAG, so `logging.getLogger('airflow.task')` would not have the same config. Is there any way to propagate Airflow's log level into KPO? Perhaps with context variables (for `@task.kubernetes` decorator) or templates?

**Comment by pykenny** [comment]
Looking up logging mechanism in `KubernetesPodOperator / KubernetesJobOperator` as well recently... so for now:

 - By default, Kubernetes collects things emit from `stdout` and `stderr` output as log records, and has their own [internal log format](https://github.com/kubernetes/design-proposals-archive/blob/main/node/kubelet-cri-logging.md). These logs are line-based, and Kubernetes CLI/API does not expose full details about the log, such as source of the log ("did it come from `stdout` or `stderr`?") and grouping ("is this line part of a grouped log, or is full content on its own?").
 - When `get_logs` is enabled, `KubernetesPodOperator` simply pulls out logs from the created pod, splits them by line breaks, and then treats each line as an individual `INFO` level message. That means if the Python application run within the pod ends up with error, you'll get a bunch of `INFO` logs that are part of the error message and back tracing information.
 - By default (without configuring a logger or its parent loggers), a Python logger only prints messages issued on `WARNING` levels and above to `stderr`. Hence merely coding `logger = logging.getLogger("airflow.task"); logger.info("Hey!")` you get nothing from Kubernetes logs.
 - Operators under `cncf.kubernetes` provider does not require Airflow to be present in the created container's runtime environment. This is a nice feature actually (you can run literally everything like Go or Java program in the pod without Python and Airflow dependencies), but it also means Airflow can't take care of log serializing within the pod for users.

That said,
 - If making `stdout`/`stderr` output available on Airflow is just what you need and you're programming everything in Python, then configure loggers (adding handlers and formatters; modify log level) beforehand in the task code to ensure log messages will be streamed to `stdout` or `stderr`.
 - Log serializing requires extra work including setting up serialize/deserialize protocol on both ends, and setting up customized operator class to replace the original logging approach.
   - [`python-json-logger`](https://github.com/madzak/python-json-logger) provides an example and converts every `LogRecord` object into single-line JSON. But I'm not sure if it supports converting the JSON back to `LogRecord` object.
   - Extra care is needed on error handling. You wouldn't want an uncaught error to slip off your loggers.
 - Otherwise, you may need to play around with [Kubernetes logging](https://kubernetes.io/docs/concepts/cluster-administration/logging/) setup and try transfer logs to some external log service. Exporting logs in [ECS](https://github.com/elastic/ecs-logging) format to some Elastic stack service is one example.
